### PR TITLE
Association testing

### PR DESCRIPTION
**Pull request by prashantiA**
New changes:
- Refactoring of datastore association code
- Tests for datastore interface

Known Issues:
- Fails on valid implementations because tests are not order agnostic for the ouput (issues with equality checks when trying to fix this, will be fixing in a future PR)

### CODE REVIEW

File: capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        new ArrayList<AssociationResult>(
+            Arrays.asList(
+                new AssociationResult("hi", 12.4f, 118.0f, false),
+                new AssociationResult("test1", 1.3f, 123.4f, true)));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    AssociationDatastore association = new AssociationDatastore(datastore);
+    association.storeResults(data, "SF");
+    AssociationDatastore newAssociation = new AssociationDatastore(datastore);
+    ArrayList<AssociationResult> retrieved = newAssociation.loadPreviousResults("SF");
+    assert (isValidAssociationDatastore(retrieved));
```
Comment (jasperlee-google) [inline capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java:131]: Do we need these extra parens? For example, can we do `assert isValidAssociationDatastore(...)`. Additionally, should we add error strings to the `assert`s via `assert False : "Some error message"`.

File: capstone/src/main/java/com/google/sps/AssociationDatastore.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    for (Entity e : results.asIterable()) {
+      if (e.getProperty("association-processed") == null
+          || !((boolean) e.getProperty("association-processed"))) {
+        String message = (String) e.getProperty(COMMENT_PROPERTY);
+        try {
+          ArrayList<String> scope = mapData.getScope((String) e.getProperty(ZIPCODE));
+          comments.add(new AssociationInput(message, scope));
+          e.setProperty("association-processed", true);
+          datastore.put(e);
+        } catch (NullPointerException exception) {
```
Comment (jasperlee-google) [inline capstone/src/main/java/com/google/sps/AssociationDatastore.java:46]: Generally `NullPointerException`s like this are systemic of an underlying bug and can catch exceptions that we did not intend to (due to bugs). Can we do a check before this intended `NullPointerException` happens?

File: capstone/src/main/java/com/google/sps/AssociationDatastore.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  public ArrayList<AssociationResult> loadPreviousResults(String scope) {
+    Query query = new Query(AssociationResult.ENTITY_KIND);
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<AssociationResult> prev = new ArrayList<AssociationResult>();
+    for (Entity entity : results.asIterable()) {
+      if (!scope.equals((String) entity.getProperty("scope"))) continue;
+      String content = (String) entity.getProperty("name");
+      float weight = (float) (double) entity.getProperty("weight");
+      float score = (float) (double) entity.getProperty("score");
```
Comment (lee12jin) [inline capstone/src/main/java/com/google/sps/AssociationDatastore.java:69]: Do you need to cast this twice?

File: capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    ArrayList<AssociationResult> data =
+        new ArrayList<AssociationResult>(
+            Arrays.asList(
+                new AssociationResult("hi", 12.4f, 118.0f, false),
+                new AssociationResult("test1", 1.3f, 123.4f, true)));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    AssociationDatastore association = new AssociationDatastore(datastore);
+    association.storeResults(data, "SF");
+    ArrayList<AssociationResult> retrieved = association.loadPreviousResults("SF");
+    assert (isValidAssociationDatastore(retrieved));
```
Comment (lee12jin) [inline capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java:100]: Please use JUnit's `assertTrue` or `assertFalse`

File: capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    ArrayList<AssociationResult> data =
+        new ArrayList<AssociationResult>(
+            Arrays.asList(
+                new AssociationResult("hi", 12.4f, 118.0f, false),
+                new AssociationResult("test1", 1.3f, 123.4f, true)));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    AssociationDatastore association = new AssociationDatastore(datastore);
+    association.storeResults(data, "Bayview");
+    ArrayList<AssociationResult> retrieved = association.loadPreviousResults("SF");
+    assert (isValidAssociationDatastore(retrieved));
```
Comment (lee12jin) [inline capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java:115]: Please use JUnit's `assertTrue` or `assertFalse`

File: capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    ArrayList<AssociationResult> data =
+        new ArrayList<AssociationResult>(
+            Arrays.asList(
+                new AssociationResult("hi", 12.4f, 118.0f, false),
+                new AssociationResult("test1", 1.3f, 123.4f, true)));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    AssociationDatastore association = new AssociationDatastore(datastore);
+    association.storeResults(data, "SF");
+    ArrayList<AssociationResult> retrieved = association.loadPreviousResults("SF");
+    assert (isValidAssociationDatastore(retrieved));
```
Comment (lee12jin) [inline capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java:146]: Please use JUnit's `assertTrue` or `assertFalse`

File: capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    assertEquals(data, retrieved);
+    association.storeResults(retrieved, "SF");
+    retrieved = association.loadPreviousResults("SF");
+    assert (isValidAssociationDatastore(retrieved));
```
Comment (lee12jin) [inline capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java:150]: Please use JUnit's `assertTrue` or `assertFalse`

File: capstone/src/main/java/com/google/sps/AssociationDatastore.java (hunk 3 above)
Comment (prashantiA) [reply capstone/src/main/java/com/google/sps/AssociationDatastore.java:69]: Yes, if you don't explicitly do the double cast it complains that the object stored is not a float.

File: capstone/src/main/java/com/google/sps/AssociationDatastore.java (hunk 2 above)
Comment (prashantiA) [reply capstone/src/main/java/com/google/sps/AssociationDatastore.java:46]: I believe that the original source of this was that there were some leftover synthetic data results mixed in with actual responses which generated null pointers off of undefined properties. I can remove the try catch and instead check for null properties when I access them.

File: capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java (hunk 1 above)
Comment (prashantiA) [reply capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java]: Moved it to assertTrue with error messages!

File: capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java (hunk 4 above)
Comment (prashantiA) [reply capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java]: Fixed this and future occurrences!

File: capstone/src/main/java/com/google/sps/AssociationDatastore.java (hunk 2 above)
Comment (jasperlee-google) [reply capstone/src/main/java/com/google/sps/AssociationDatastore.java]: Thanks, that sounds like a good way to approach it.

**Pull request merged**